### PR TITLE
[CQT-254] Add instructions and directives to OpenSquirrel

### PR DESCRIPTION
**Pull request by elenbaasc**
Instructions added to OpenSquirrel:
- `init`
- `SWAP` (with Matrix semantic)
- `wait`

Further changes:
- Refactor of functionality to get expanded arguments in Parser.
- Small refactor of `visit_non_unitary` of `_WriterImpl` in `writer.py`.
- Renamed qubit arguments in default instructions from `q` to `qubit`.
- Added tests for the instructions (also as part of existing tests).

### CODE REVIEW

File: opensquirrel/parser/libqasm/parser.py [hunk 1]
```diff
@@ -94,8 +94,10 @@ def _get_bits(
         return ret
 
     @classmethod
-    def _get_gate_operands(
-        cls, instruction: cqasm.semantic.GateInstruction, register_manager: RegisterManager
+    def _get_instruction_operands(
+        cls,
+        instruction: cqasm.semantic.GateInstruction | cqasm.semantic.NonGateInstruction,
```
Comment (rturrado) [inline opensquirrel/parser/libqasm/parser.py:99]: Can we write?

```suggestion
        instruction: cqasm.semantic.Instruction,
```

File: opensquirrel/writer/writer.py [hunk 2]
```diff
@@ -50,20 +50,23 @@ def visit_float(self, f: SupportsFloat) -> str:
...
-            qubit_argument = non_unitary.arguments[0].accept(self)  # type: ignore[index]
-            self.output += f"{bit_argument} = {non_unitary.name} {qubit_argument}\n"
+            non_unitary_name = f"{bit_argument} = {non_unitary_name}"
         else:
-            qubit_argument = non_unitary.arguments[0].accept(self)  # type: ignore[index]
-            self.output += f"{non_unitary.name} {qubit_argument}\n"
+            params = [param.accept(self) for param in non_unitary.arguments[1::]]  # type: ignore[index]
+            if params:
+                non_unitary_name = f"{non_unitary_name}({', '.join(params)})"
+        self.output += f"{non_unitary_name} {qubit_argument}\n"
```
Comment (rturrado) [inline opensquirrel/writer/writer.py:62]: Suggestion  (not tested):

```
        bit_argument = non_unitary.arguments[1].accept(self)  if non_unitary.name == "measure" else ""  # type: ignore[index]
        bit_section = f"{bit_argument} = " if not bit_argument.empty() elsse ""
        params = [param.accept(self) for param in non_unitary.arguments[1::]]  # type: ignore[index]
        param_section = f"({', '.join(params)})" if params else  ""
        qubit_argument = non_unitary.arguments[0].accept(self)  # type: ignore[index]
        self.output += f"{bit_section}{non_unitary.name}{param_section} {qubit_argument}\n"
```

**Review by rturrado — COMMENTED** [review]
Beautiful! Good work!

File: test/test_wait.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
+from opensquirrel import Circuit, CircuitBuilder
+from opensquirrel.ir import Bit
+
+
+def test_wait() -> None:
```
Comment (juanboschero) [inline test/test_wait.py:5]: I would be interested in a test utilizing a multiqubit gate, barriers and waits to see if the logic flows correctly.

**Review by juanboschero — CHANGES_REQUESTED** [review]

File: opensquirrel/ir.py [hunk 4]
```diff
@@ -390,6 +396,34 @@ def get_qubit_operands(self) -> list[Qubit]:
...
+        NonUnitary.__init__(self, generator, arguments)
+        self.qubit = Qubit(qubit)
+
+    def __repr__(self) -> str:
+        return f"Init(qubit={self.qubit})"
+
+    @property
+    def is_abstract(self) -> bool:
+        return self.arguments is None
+
```
[the remark this answers is not in the document]
Comment (elenbaasc) [reply opensquirrel/ir.py:415]: No, don't think so. I will also remove it from the `Reset` semantic.

File: opensquirrel/parser/libqasm/parser.py [hunk 5]
```diff
@@ -215,13 +205,13 @@ def circuit_from_string(self, s: str) -> Circuit:
...
                 instruction_generator = self._get_gate_f(statement)
-                expanded_args = Parser._get_expanded_gate_args(statement, register_manager)
+                expanded_args = Parser._get_expanded_instruction_args(statement, register_manager)
             elif Parser._is_non_unitary_instruction(statement):
                 instruction_generator = self._get_non_unitary_f(statement)
                 if statement.name == "measure":
                     expanded_args = Parser._get_expanded_measure_args(statement.operands, register_manager)
                 else:
-                    expanded_args = Parser._get_expanded_reset_args(statement.operands, register_manager)
+                    expanded_args = Parser._get_expanded_instruction_args(statement, register_manager)
```
[the remark this answers is not in the document]
Comment (elenbaasc) [reply opensquirrel/parser/libqasm/parser.py:214]: Like this then:
```
instruction_generator = self._get_gate_f(statement) if Parser._is_gate_instruction(statement) \
                else self._get_non_unitary_f(statement)
            expanded_args = Parser._get_expanded_instruction_args(statement, register_manager) \
                if statement.name != "measure" \
                else Parser._get_expanded_measure_args(statement.operands, register_manager)
```

[the remark this answers is not in the document]
Comment (elenbaasc) [reply opensquirrel/parser/libqasm/parser.py:214]: Actually, this does not work because a GateInstruction does not have a 'name' attribute.

File: opensquirrel/writer/writer.py (hunk 2 above)
Comment (elenbaasc) [reply opensquirrel/writer/writer.py:62]: It's nice to generalize the generation of these statements. However, this suggestion does not work because the 1-th argument of a measure statement is the bit value, which is then also interpreted as a parameter.

We will need to revisit this code when we include the support for parameterized non-unitaries (with axis parameters)

File: test/exporter/test_cqasmv1_exporter.py [hunk 6]
```diff
@@ -144,15 +170,15 @@ def test_reset() -> None:
     )
 
 
-def test_all_supported_gates() -> None:
+def test_all_instructions() -> None:
```
[the remark this answers is not in the document]
Comment (elenbaasc) [reply test/exporter/test_cqasmv1_exporter.py:173]: Not only unitary instructions are tested.

File: test/test_wait.py (hunk 3 above)
Comment (elenbaasc) [reply test/test_wait.py:5]: Any suggestions?

File: opensquirrel/writer/writer.py (hunk 2 above)
Comment (rturrado) [reply opensquirrel/writer/writer.py:62]: OK, true, thanks. We could check for `non_unitary_name != "measure"` in the `params =` line, but that is overcomplicating the code.

What I would do then is to set `self.output` in both blocks, mainly to avoid that line 57 (`non_unitary_name` is not really a name there, is a bit argument plus an equals sign plus a name).

File: test/exporter/test_cqasmv1_exporter.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (rturrado) [reply test/exporter/test_cqasmv1_exporter.py:173]: I think so. That's the goal of the test. I know it includes a couple of `measure` statements. I would say:
- either test all instructions then (measure, reset, init...)
- or just test all unitary instructions and remove the `measure` statements.

[the remark this answers is not in the document]
Comment (elenbaasc) [reply test/exporter/test_cqasmv1_exporter.py:173]: Good point! I've added the missing instructions.

File: test/test_init.py [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+        version 3.0
+
+        qubit[4] q
+
+        H q[0]
+        H q[1:2]
+        init q[2:3]
+        H q[3]
+        init q[0:1]
+        """,
```
Comment (rturrado) [inline test/test_init.py:47]: Same here.
If we want to test `init` in the middle of a circuit, I would put it in a separate test.
BTW, I would propose to write this set of tests as a parametrized test. You can write the programs and the outputs in a single line. Guy wrote recently one like that (`test_simplest` in test_libqasm.py). E.g.:

```
    (
        "version 3.0; qubit[4] q; init q[2:3]",
        "version 3.0\n\nqubit[4] q\n\ninit q[2]\ninitq[3]",
    )
```

```suggestion
        version 3.0

        qubit[4] q

        init q[2:3]
        """,
```

File: test/test_wait.py [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+    qc = Circuit.from_string(
```
Comment (rturrado) [inline test/test_wait.py:6]: Same comments from `init`:
- Remove extra instructions, keep tests just to the minimum needed.
- Use parametrized tests.

**Review by rturrado — COMMENTED** [review]
Some more comments, mainly regarding to tests.

File: test/test_circuit_builder.py [hunk 9]
```diff
@@ -21,39 +21,27 @@ def test_simple(self) -> None:
...
+        ("builder", "expected_result"),
+        [
+            (CircuitBuilder(2, 2).I(0).X(0).Y(0).Z(0), [I(0), X(0), Y(0), Z(0)]),
+            (CircuitBuilder(2, 2).Rx(0, -1).Ry(1, 1).Rz(0, 3.14), [Rx(0, -1), Ry(1, 1), Rz(0, 3.14)]),
+            (CircuitBuilder(2, 2).CZ(0, 1).CNOT(1, 0).SWAP(0, 1), [CZ(0, 1), CNOT(1, 0), SWAP(0, 1)]),
+            (CircuitBuilder(2, 2).measure(0, 0).measure(1, 1), [Measure(0, 0), Measure(1, 1)]),
+            (CircuitBuilder(2, 2).init(0).init(1), [Init(0), Init(1)]),
+            (CircuitBuilder(2, 2).reset(0).reset(1), [Reset(0), Reset(1)]),
+            (CircuitBuilder(2, 2).barrier(0).barrier(1), [Barrier(0), Barrier(1)]),
+            (CircuitBuilder(2, 2).wait(0, 1).wait(1, 2), [Wait(0, 1), Wait(1, 2)]),
```
Comment (rturrado) [inline test/test_circuit_builder.py:34]: What do you think it is more legible? I leave it up to you. I like both forms.

```suggestion
            (
                CircuitBuilder(2, 2).I(0).X(0).Y(0).Z(0),
                [I(0), X(0), Y(0), Z(0)]
            ),
            (
                CircuitBuilder(2, 2).Rx(0, -1).Ry(1, 1).Rz(0, 3.14),
                [Rx(0, -1), Ry(1, 1), Rz(0, 3.14)]
            ),
            (
                CircuitBuilder(2, 2).CZ(0, 1).CNOT(1, 0).SWAP(0, 1),
                [CZ(0, 1), CNOT(1, 0), SWAP(0, 1)]
            ),
            (
                CircuitBuilder(2, 2).measure(0, 0).measure(1, 1),
                [Measure(0, 0), Measure(1, 1)]
            ),
            (
                CircuitBuilder(2, 2).init(0).init(1),
                [Init(0), Init(1)]
            ),
            (
                CircuitBuilder(2, 2).reset(0).reset(1),
                [Reset(0), Reset(1)]
            ),
            (
                CircuitBuilder(2, 2).barrier(0).barrier(1),
                [Barrier(0), Barrier(1)]
            ),
            (
                CircuitBuilder(2, 2).wait(0, 1).wait(1, 2),
                [Wait(0, 1), Wait(1, 2)]
            ),
```

File: test/test_circuit_builder.py [hunk 10]
```diff
@@ -21,39 +21,27 @@ def test_simple(self) -> None:
             CNOT(0, 1),
         ]
 
-    def test_identity(self) -> None:
```
Comment (rturrado) [inline test/test_circuit_builder.py:24]: What happened to these removed tests?

File: test/test_init.py [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+
+
+@pytest.mark.parametrize(
+    ("cqasm_string", "expected_result"),
+    [
+        ("version 3.0; qubit[2] q; init q[1]; init q[0]", "version 3.0\n\nqubit[2] q\n\ninit q[1]\ninit q[0]\n"),
+        (
+            "version 3.0; qubit[4] q; init q[2:3]; init q[1, 0]",
+            "version 3.0\n\nqubit[4] q\n\ninit q[2]\ninit q[3]\ninit q[1]\ninit q[0]\n",
+        ),
```
Comment (rturrado) [inline test/test_init.py:14]: In this case, I really think two lines help.
And would you prefer to write the output program in several lines, I leave that up to you as well.

```suggestion
        (
            "version 3.0; qubit[2] q; init q[1]; init q[0]",
            "version 3.0\n\nqubit[2] q\n\ninit q[1]\ninit q[0]\n"
        ),
        (
            "version 3.0; qubit[4] q; init q[2:3]; init q[1, 0]",
            "version 3.0\n\nqubit[4] q\n\ninit q[2]\ninit q[3]\ninit q[1]\ninit q[0]\n",
        ),
```

File: test/test_reset.py [hunk 12]
```diff
@@ -1,89 +1,29 @@
...
-"""
-    )
+@pytest.mark.parametrize(
+    ("cqasm_string", "expected_result"),
+    [
+        ("version 3.0; qubit[2] q; reset q[1]; reset q[0]", "version 3.0\n\nqubit[2] q\n\nreset q[1]\nreset q[0]\n"),
+        (
+            "version 3.0; qubit[4] q; reset q[2:3]; reset q[1, 0]",
+            "version 3.0\n\nqubit[4] q\n\nreset q[2]\nreset q[3]\nreset q[1]\nreset q[0]\n",
+        ),
```
Comment (rturrado) [inline test/test_reset.py:14]: Same same but different.

```suggestion
        (
            "version 3.0; qubit[2] q; reset q[1]; reset q[0]",
            "version 3.0\n\nqubit[2] q\n\nreset q[1]\nreset q[0]\n"
        ),
        (
            "version 3.0; qubit[4] q; reset q[2:3]; reset q[1, 0]",
            "version 3.0\n\nqubit[4] q\n\nreset q[2]\nreset q[3]\nreset q[1]\nreset q[0]\n",
        ),
```

**Review by rturrado — CHANGES_REQUESTED** [review]
Many thanks for the commits.

Almost there I think!

File: test/test_wait.py (hunk 3 above)
Comment (juanboschero) [reply test/test_wait.py]:
```
builder = CircuitBuilder(3)
  builder.H(0)
  builder.H(1)
  builder.wait(0, 1)
  builder.barrier(1)
  builder.wait(3,1)
  builder.barrier(0)
  builder.CNOT(0, 1)
  builder.wait(0, 3)
```

File: opensquirrel/parser/libqasm/parser.py [hunk 13]
```diff
@@ -120,17 +122,22 @@ def _get_named_gate_param(cls, gate: cqasm.semantic.Gate) -> Any:
         return cls._ast_literal_to_ir_literal(gate.parameter)
 
     @classmethod
-    def _get_expanded_gate_args(
-        cls, instruction: cqasm.semantic.GateInstruction, register_manager: RegisterManager
+    def _get_expanded_instruction_args(
+        cls,
+        instruction: cqasm.semantic.GateInstruction | cqasm.semantic.NonGateInstruction,
```
[the remark this answers is not in the document]
Comment (elenbaasc) [reply opensquirrel/parser/libqasm/parser.py:127]: Yes, I had changed that... strange

File: test/test_circuit_builder.py (hunk 10 above)
Comment (elenbaasc) [reply test/test_circuit_builder.py:24]: It has been added to the `test_instructions` as part of the first test `"pauli gates"`.

File: opensquirrel/parser/libqasm/parser.py [hunk 14]
```diff
@@ -94,8 +94,10 @@ def _get_bits(
...
     @classmethod
-    def _get_gate_operands(
-        cls, instruction: cqasm.semantic.GateInstruction, register_manager: RegisterManager
+    def _get_instruction_operands(
+        cls,
+        instruction: cqasm.semantic.Instruction,
+        register_manager: RegisterManager,
     ) -> list[list[Any]]:
         """Get the list of lists of operands of a gate.
         Notice that a gate just has a list of operands. The outer list is needed to support SGMQ.
```
[the remark this answers is not in the document]
Comment (elenbaasc) [reply opensquirrel/parser/libqasm/parser.py:103]: I choose the first option.

File: test/test_reset.py (hunk 12 above)
Comment (elenbaasc) [reply test/test_reset.py]: I accepted your changes, but `ruff format` prescribes the old form...

File: test/test_init.py (hunk 11 above)
Comment (elenbaasc) [reply test/test_init.py]: I accepted your changes, but `ruff format` prescribes the old form...

File: test/test_circuit_builder.py [hunk 15]
```diff
@@ -21,39 +21,27 @@ def test_simple(self) -> None:
...
+            (CircuitBuilder(2, 2).I(0).X(0).Y(0).Z(0), [I(0), X(0), Y(0), Z(0)]),
+            (CircuitBuilder(2, 2).Rx(0, -1).Ry(1, 1).Rz(0, 3.14), [Rx(0, -1), Ry(1, 1), Rz(0, 3.14)]),
+            (CircuitBuilder(2, 2).CZ(0, 1).CNOT(1, 0).SWAP(0, 1), [CZ(0, 1), CNOT(1, 0), SWAP(0, 1)]),
+            (CircuitBuilder(2, 2).measure(0, 0).measure(1, 1), [Measure(0, 0), Measure(1, 1)]),
+            (CircuitBuilder(2, 2).init(0).init(1), [Init(0), Init(1)]),
+            (CircuitBuilder(2, 2).reset(0).reset(1), [Reset(0), Reset(1)]),
+            (CircuitBuilder(2, 2).barrier(0).barrier(1), [Barrier(0), Barrier(1)]),
+            (CircuitBuilder(2, 2).wait(0, 1).wait(1, 2), [Wait(0, 1), Wait(1, 2)]),
+        ],
+        ids=["pauli gates", "rotation gates", "two-qubit gates", "measure", "init", "reset", "barrier", "wait"],
```
Comment (rturrado) [inline test/test_circuit_builder.py:36]: You can write almost anything as a test ID. And I understand we don't have a convention. But if you do a general search across the codebase you could see that the pattern we follow is:

- Uppercase allowed.
- Use underscore instead of spaces.

```suggestion
        ids=["Pauli_gates", "rotation_gates", "two-qubit_gates", "measure", "init", "reset", "barrier", "wait"],
```

File: test/test_init.py (hunk 11 above)
Comment (rturrado) [reply test/test_init.py:14]: Maybe worth using `# fmt: skip`?

File: test/test_reset.py (hunk 12 above)
Comment (rturrado) [reply test/test_reset.py:14]: Maybe worth using `# fmt: skip`?

File: test/test_circuit_builder.py (hunk 9 above)
Comment (rturrado) [reply test/test_circuit_builder.py:34]: Maybe worth using `# fmt: skip`?

File: test/test_circuit_builder.py (hunk 10 above)
Comment (rturrado) [reply test/test_circuit_builder.py:24]: OK.

**Review by rturrado — APPROVED** [review]
LGTM!

Thanks for the work and the patience!

**Comment by elenbaasc** [comment]
Thanks for the thorough review! Very much appreciated!

**Review by juanboschero — APPROVED** [review]
